packages/common/src/ui/layout/tables/RemoteDataTable.tsx

```diff
@@ -0,0 +1,179 @@
+/* eslint-disable react/jsx-key */
+import React, { ReactNode, useEffect, useState } from 'react';
+
+import {
+  Column,
+  ColumnInstance,
+  Row,
+  usePagination,
+  useSortBy,
+  useTable,
+} from 'react-table';
+import { QueryObserverResult } from 'react-query';
+import clsx from 'clsx';
+
+import {
+  Box,
+  CircularProgress,
+  Grid,
+  makeStyles,
+  TableBody,
+  TableCell,
+  TableHead,
+  TableRow,
+  TableContainer,
+  TablePagination,
+  Table as MuiTable,
+} from '@material-ui/core';
+
+import { SortAsc, SortDesc } from '../../icons';
+import { DEFAULT_PAGE_SIZE } from '.';
+
+export interface QueryProps {
+  first: number;
+  offset: number;
+  sort?: string;
+  desc?: boolean;
+}
+
+export interface QueryResponse<T> {
+  data: T[];
+  totalLength: number;
+}
+
+const useStyles = makeStyles(theme => ({
+  container: { marginBottom: 100 },
+  bodyCell: {
+    padding: 9.5,
+  },
+  clickable: {
+    cursor: 'pointer',
+  },
+  headerCell: {
+    ...theme.typography.th,
+  },
+  loadingIndicator: { marginLeft: 'auto', marginRight: 'auto' },
+  loadingIndicatorContainer: { display: 'flex', marginTop: 50 },
+}));
+
+interface TableProps<T extends Record<string, unknown>> {
+  columns: Column<T>[];
+  onFetchData: (props: QueryProps) => Promise<QueryObserverResult<T>>;
+  onRowClick?: <T extends Record<string, unknown>>(row: Row<T>) => void;
+}
+
+const renderSortIcon = <D extends Record<string, unknown>>(
+  column: ColumnInstance<D>
+) => {
+  if (!column.isSorted) return null;
+
+  return !!column.isSortedDesc ? <SortDesc /> : <SortAsc />;
+};
+
+export const RemoteDataTable = <T extends Record<string, unknown>>({
+  columns,
+  onFetchData,
+  onRowClick,
+}: TableProps<T> & { children?: ReactNode }): JSX.Element => {
+  const classes = useStyles();
+  const [pageSize, setPageSize] = useState(DEFAULT_PAGE_SIZE);
```

Comment:
I'm not happy with a lot of `useState` calls here but like it slightly better than mutating an object